### PR TITLE
[#40] Inspections for partial functions on infinite lists

### LINKED ISSUE #40: Inspections for partial functions on infinite lists
- [x] `length`
- [x] `reverse`
- [x] `isSuffixOf`
- [x] `sum`
- [x] `product`
- [x] `genericLength`

### PR DESCRIPTION
**Pull request by vrom911**
Resolves #40

### CODE REVIEW

File: src/Stan/Analysis/Analyser.hs [hunk 1]
```diff
@@ -56,17 +58,20 @@ analyseNameMeta insId nameMeta hie@HieFile{..} =
...
         -> (Identifier, IdentifierDetails TypeIndex)
         -> Maybe RealSrcSpan
-    findHeadUsage srcSpan (identifier, details) = do
+    findUsage typeIxs srcSpan (identifier, details) = do
         Right name <- Just identifier
         guard $ Set.notMember (IEThing Import) $ identInfo details
 
         guard $ compareNames nameMeta name
 
+        guard $ any isJust $ map (\i -> hieMatchType i pat hie_types) typeIxs
```
Comment (chshersh) [inline src/Stan/Analysis/Analyser.hs:75]: Would it be better to have a single `guard` with multiple cases? Like:

```haskell
guard
    $  Set.nonMember ...
    && compareNames ...
    && ...
```

File: src/Stan/Inspection/Infinite.hs [hunk 2]
```diff
@@ -35,29 +42,44 @@ infiniteInspectionsMap = fromList $ map (mapToFst inspectionId)
...
     , inspectionCategory = Category.infinite :| [Category.list]
     , inspectionSeverity = Warning
-    , inspectionAnalysis = FindName nameMeta
+    , inspectionAnalysis = FindName nameMeta pat
     }
 
 -- | 'Inspection' for 'stan0101' — infinite 'reverse' @STAN-0101@.
 stan0101 :: Inspection
-stan0101 = mkInfiniteInspection (Id "STAN-0101") (mkBaseListMeta "reverse")
+stan0101 = mkInfiniteInspection (Id "STAN-0101") (mkBaseListMeta "reverse") PatternAnything
```
Comment (chshersh) [inline src/Stan/Inspection/Infinite.hs:65]: In such cases I with Haskell supported default arguments...

File: src/Stan/Inspection/Infinite.hs [hunk 3]
```diff
@@ -35,29 +42,44 @@ infiniteInspectionsMap = fromList $ map (mapToFst inspectionId)
...
+stan0104 :: Inspection
+stan0104 = mkInfiniteInspection (Id "STAN-0104") (mkBaseOldListMeta "genericLength") PatternAnything
+
+-- | 'Inspection' for 'stan0105' — infinite 'Data.Foldable.sum' @STAN-0105@.
+stan0105 :: Inspection
+stan0105 = mkInfiniteInspection (Id "STAN-0105") (mkBaseFoldableMeta "sum") (PatternFun patternList PatternAnything)
+
+-- | 'Inspection' for 'stan0106' — infinite 'Data.Foldable.product' @STAN-0106@.
+stan0106 :: Inspection
+stan0106 = mkInfiniteInspection (Id "STAN-0106") (mkBaseFoldableMeta "product") (PatternFun patternList PatternAnything)
```
Comment (chshersh) [inline src/Stan/Inspection/Infinite.hs:85]: This pattern seems common enough. Should we create a variable for it, like `listFunPattern`?

File: src/Stan/Inspection/Infinite.hs (hunk 2 above)
Comment (vrom911) [reply src/Stan/Inspection/Infinite.hs:65]: :crying_cat_face:

File: src/Stan/Hie/Match.hs [hunk 4]
```diff
@@ -66,15 +66,12 @@ data Pattern
 
 {- | Matching function that searches the array of types recursively.
 -}
-hieMatchType
+hieMatchPattern
     :: TypeIndex   -- ^ Index of the current expression type
     -> Pattern  -- ^ Our search query
     -> Array TypeIndex HieTypeFlat  -- ^ Array of all types in HIE file
```
Comment (chshersh) [inline src/Stan/Hie/Match.hs:72]: Here is just a thought I had: what if we reorder arguments in this function to the following order:

```haskell
hieMatchPattern :: Array TypeIndex -> Pattern -> TypeIndex -> Bool
```

With the new order you whould be able to use partially applied `hieMatchPattern` and rewrite things like `any (\i -> hieMatchPattern i pat hie_types)` to `any (hieMatchPattern hie_types pat)`. Also, probably, with some extra internal functions, you can avoid passing `arr` as an argument inside internal recursive functions, but that's a minor suggestion, maybe not worth it.

File: src/Stan/Hie/Match.hs [hunk 5]
```diff
@@ -89,12 +86,12 @@ hieMatchType i pat arr =
...
+        && checkWith (\(_, ix) a -> hieMatchPattern ix a arr) hieArgs args
     satisfyPattern (HFunTy i1 i2) (PatternFun p1 p2) =
-           isJust (hieMatchType i1 p1 arr)
-        && isJust (hieMatchType i2 p2 arr)
-    satisfyPattern (HQualTy _ ix) p = isJust $ hieMatchType ix p arr
-    satisfyPattern (HForAllTy _ ix) p = isJust $ hieMatchType ix p arr
+           hieMatchPattern i1 p1 arr
+        && hieMatchPattern i2 p2 arr
+    satisfyPattern (HQualTy _ ix) p = hieMatchPattern ix p arr
+    satisfyPattern (HForAllTy _ ix) p = hieMatchPattern ix p arr
```
Comment (chshersh) [inline src/Stan/Hie/Match.hs]: I like the `Bool` simplification! Made code much simpler 👍

File: src/Stan/Analysis/Analyser.hs [hunk 6]
```diff
@@ -56,17 +58,22 @@ analyseNameMeta insId nameMeta hie@HieFile{..} =
...
-        guard $ Set.notMember (IEThing Import) $ identInfo details
-
-        guard $ compareNames nameMeta name
+        guard
+            -- not in the imports
+            $ Set.notMember (IEThing Import) (identInfo details)
+            -- exact name/module/package
+            && compareNames nameMeta name
+            -- compatible with the given pattern
+            && any (hieMatchPattern hie_types pat) typeIxs
```
Comment (chshersh) [inline src/Stan/Analysis/Analyser.hs:77]: This is genius! 😍

**Pull request merged**